### ISSUE TITLE
Use default slots to save space

### ISSUE THREAD

**Issue by timothyis** [EmilStenstrom/django-components#231]
Related to #189, I think it would be great to have a "default" slot. Then you can fill a component_block without using the slot syntax if a component only has one slot and save a lot of lines. E.g., 

```django
{% component_block "box" %}
	<h1>hello</h1>
{% endcomponent_block %}
```

with the component HTML:
```django
<div>
	{% slot default %}
	{% endslot %}
</div>
```

This would help with a lot of components that are nested and become messy. For example, we have a table component:
```django
{% component_block "table" columns="sm|default (md)|lg (todo)" only %}
	{% slot "rows" %}
		{% component_block "table-row" classes=classes %}
			{% slot "columns" %}
				{% component_block "table-column" %}
					{% slot value %}
						<span>Some value</span>
					{% endslot %}
				{% endcomponent_block %}
				{% component_block "table-column" %}
					{% slot value %}
						<span>Some value</span>
					{% endslot %}
				{% endcomponent_block %}
			{% endslot %}
		{% endcomponent_block %}
	{% endslot %}
{% endcomponent_block %}
```

Would become:
```django
{% component_block "table" columns="sm|default (md)|lg (todo)" only %}
	{% component_block "table-row" classes=classes %}
		{% component_block "table-column" %}
			<span>Some value</span>
		{% endcomponent_block %}
		{% component_block "table-column" %}
			<span>Some value</span>
		{% endcomponent_block %}
	{% endcomponent_block %}
{% endcomponent_block %}
```

You can imagine the amount of space this will save when it has more rows/columns too.

This model also follows closely with how frontend JS frameworks work (without slots), so it will be easier for those of us starting there to understand.

**Comment by EmilStenstrom** [comment]
I think this makes a lot of sense, and would be open to a PR to add this. 

Your thoughts @lemontheme?

**Comment by lemontheme** [comment]
@timothyis is right about other JS frameworks doing it this way. If we _really_ want to follow their example we should make it optional for slots to specify a name _if and only if_ there is only one slot the template. (I think that's preferable to having a semi-magical 'default' label.)

For example, the component template would look like this:

```
<div>
	{% slot %}  {# look Ma, no name  #}
	{% endslot %}
</div>
```

I can more or less see how we'd make this work on the implementation side. :)

**Comment by EmilStenstrom** [comment]
I like the idea of using concepts from JS frameworks if that makes the code clearer. But in this case I'm not sure it does. 

Say you see a component call like this:
```django
{% component "something" %}
    Content here
{% endcomponent %}
```

And go look in the template so see how that component works:
```django
{% slot "header" %}{% endslot %}
{% slot %}{% endslot %}
{% slot "footer" %}{% endslot %}
```

Compared to this:
```django
{% slot "header" %}{% endslot %}
{% slot "body" default %}{% endslot %}
{% slot "footer" %}{% endslot %}
```

I think the latter makes things easier to understand just by looking at the code.

**Comment by timothyis** [comment]
I definitely think a merge between the two worlds is the best way to do things, rather than taking complete inspiration (coming from someone who is primarily a React dev 😅).

`default` works in this case since Django templates seem to be very much based in context whereas in React, you don't use slots at all, but just multiple components that are scoped to a wrapping component, which provides the context.

I'd say if this were to really take inspiration of JS frameworks, there'd be no slots and only guidance on using components within other components. See radix-ui for example: https://www.radix-ui.com/docs/primitives/components/dropdown-menu (index.jsx component usage). But I'm not sure this is the goal here, even though I'm a fan of this in some cases and already use multiple components together to create one (tabs for example: "tab-container" and "tab-item"). However, it does make it _more_ complicated in various scenarios. 

Just spewing out thoughts here, hoping they're helpful.

**Comment by lemontheme** [comment]
@EmilStenstrom – So, I was taking a page out of Vue's book, where default slots are permitted __if and only if__ the component defines a __single__ slot. There, the situation you describe simply can't arise.

A `default` modifier might make sense too. Beware though that there has been discussion about other slot modifiers, e.g. `optional`, so we should ask ourselves how many extra special keywords we're willing to take on. My answer would almost always be fewer = better.

**Comment by EmilStenstrom** [comment]
@lemontheme Since we're back to supporting a version of #232 that would work with a default slot I think we could look at this again. I think a default keyword makes sense, and requiring it to be able to use the default slot mechanism. Making this work when there's just a single slot makes things to magical, and also leaves no trace in the slot definition which I don't like.

**Comment by lemontheme** [comment]
I get the argument for being more explicit. To me it does appear to run counter to one of the benefits of default slots: reducing typing work/template verbosity.

See, I reason that in many, many cases, components will only have a single slot. At that point, why go to the effort of finding a name for this slot? A name is just an identifier to differentiate one slot from another, after all. So a nameless slot gets the job done with less typing and less thinking. Now, on the filling end of things: again, a slot with no name takes a fill with no name. But if there's only one target slot, why use a fill tag at all, since, again, there are no other fill tags to differentiate between. Might as well dump the filling content directly in the span of the component_block pair.

Let's compare our two design proposals:

#### Yours

Component template:

```django
<div>
  {% slot "name" default required %}
  {% endslot %}
</div>
```

Using the component


```django
{% component_block "single_slot_comp" %}
<p>hi there!</p>
{% endcomponent_block %}
```

Or

```django
{% component_block "single_slot_comp" %}
  {% fill <name>(optional) %} <p>hi there!</p> {% endfill %}
{% endcomponent_block %}
```

#### Mine (stolen from Vue.js)

Component template:

```django
<div>
  {% slot required %}
  {% endslot %}
</div>
```

Using the component: (Same as before)

```django
{% component_block "single_slot_comp" %}
<p>hi there!</p>
{% endcomponent_block %}
```

Or 

```django
{% component_block "single_slot_comp" %}
  {% fill %} <p>hi there!</p> {% endfill %}
{% endcomponent_block %}
```

----

The benefit of your approach is that you can have a default slot _in addition to_ standard slots. The downside is users need to remember a new keyword (something I'm guilty of as well in the latest PR, so no stones being thrown here). =p

The benefit of my approach (again, stolen from Vue) is that, because it's an 'either-or' thing (either a single nameless slot or multiple named slots), a simple, single-slot component take less typing to define. The downside is that users need to learn a new 'convention'. (Which in this case is mitigated by the fact that this convention occurs in other frameworks, making django-components feel more familiar for JS refugees.)

**Comment by lemontheme** [comment]
TL;DR Should users be required to learn a new keyword ('default') or a new convention ('solitary slots can be nameless, in which case they can be used as a default slot')?

@timothyis, would love to hear your thoughts!

**Comment by EmilStenstrom** [comment]
Going back to the Zen of Python, we have have [explicit is better than implicit](https://towardsdatascience.com/the-zen-of-python-a-guide-to-pythons-design-principles-93f3f76d088a#ac0c). This is a big Python vs JS difference! And it does run in contrary to the goal of saving keystrokes for people writing components.

But this ticket is not about people **making** components though, but for people **using** them. So while you write a component once, you might be using it 10 or 100 times in a project. And as long as the intent is clear when using a component (and I think the original suggestion is!) I think saving keystrokes when using a component is a worthy cause (this is also the reason for #232). 

So naming slots when you define them has two more purposes imo:
* Documenting the reason why there's a slot there in the most basic way possible: Giving it a name. 
* Making the jump from one to two slots easier. Why should I have to make up a name for the first slot to add a second one?

TLDR: In not in favour of having nameless slots in templates.

Another note: The only reason the "default" keyword is needed is to support filling a default slot when there's more than one slot defined. If we skip that case, and optimize for the common case with just having one slot, the default keyword is not needed. A downside of that is that you break all call sites for a component when adding a second slot, but again, multiple slots might be unusual?

**Comment by EmilStenstrom** [comment]
Yay, we have a path forward! 

Yes, that's the design I'm thinking of for using a component. And for defining a component we use `{% slot "name" default %}` ?

**Comment by lemontheme** [comment]
Indeed, progress!

Yup, let's go with the new `default` keyword.

I can get started on this fairly quickly. Only question now is what to tackle first: this or #232?

**Comment by EmilStenstrom** [comment]
I think this one is simpler? So this one first, if I can pick from the sidelines! :)

**Comment by lemontheme** [comment]
You got it!

**Comment by lemontheme** [comment]
Btw – slightly off-topic – what would you think of us setting up one of those GH projects to track our work? Might be nice for others to see what we're working on/the general timeline?

**Comment by timothyis** [comment]
Apologies, I've been extremely busy as of late! I'm very happy with the outcome here 🎉

**Comment by timothyis** [comment]
> Btw – slightly off-topic – what would you think of us setting up one of those GH projects to track our work? Might be nice for others to see what we're working on/the general timeline?

Just to note that we use this for work and they are fantastic for keeping projects transparent to users and for keeping contributors organised.

**Comment by EmilStenstrom** [comment]
About GitHub projects: I think they are great, but as long as we have one project going at a time, and about 10 issues in total, I’m not sure there’s much benefit.

**Comment by lemontheme** [comment]
Yeah, that's fair. Besides, I suppose the 'Version 1.0' milestone serves the same purpose of communicating to others what's on the roadmap.
